### ISSUE TITLE
Can not publish Topics of custom messages like imgmsg(CvBridge ) and  pcl2(sensor_msgs_py) using 'ros2 bag play'

### ISSUE THREAD

**Issue by jaan1729** [ros2/rosbag2#655]
## Description

I built ros2 rolling from source to use the new packages like rosbag2_py and sensor_messages_py.

I referred [kitti2bag](https://github.com/tomas789/kitti2bag/blob/master/kitti2bag/kitti2bag.py) repo and rosbag2 [test example](https://github.com/ros2/rosbag2/blob/master/rosbag2_py/test/test_sequential_writer.py) to record rosbags for Kitti dataset. 
Here is the core program I used.

```python
import os
import sys
import rclpy as rospy
from rclpy.serialization import deserialize_message, serialize_message
from rosidl_runtime_py.utilities import get_message
from std_msgs.msg import String
import transformations

if os.environ.get('ROSBAG2_PY_TEST_WITH_RTLD_GLOBAL', None) is not None:
    # This is needed on Linux when compiling with clang/libc++.
    # TL;DR This makes class_loader work when using a python extension compiled with libc++.
    #
    # For the fun RTTI ABI details, see https://whatofhow.wordpress.com/2015/03/17/odr-rtti-dso/.
    sys.setdlopenflags(os.RTLD_GLOBAL | os.RTLD_LAZY)

from .common import get_rosbag_options  # noqa
import rosbag2_py  # noqa

import tf2_ros as tf
import os
import cv2


import rosbag2_py
import progressbar
from tf2_msgs.msg import TFMessage
from datetime import datetime
from std_msgs.msg import Header
from sensor_msgs.msg import CameraInfo, Imu, PointField, NavSatFix
import sensor_msgs_py.point_cloud2 as pcl2
from geometry_msgs.msg import TransformStamped, TwistStamped, Transform
from cv_bridge import CvBridge
import numpy as np
import argparse
import pykitti

def create_topic(writer, topic_name, topic_type, serialization_format='cdr'):
    """
    Create a new topic.
    :param writer: writer instance
    :param topic_name:
    :param topic_type:
    :param serialization_format:
    :return:
    """
    topic_name = topic_name
    topic = rosbag2_py.TopicMetadata(name=topic_name, type=topic_type,
                                     serialization_format=serialization_format)

    writer.create_topic(topic)



def save_imu_data(bag, kitti, imu_frame_id, topic):
    print("Exporting IMU")
    msg_topic = topic
    create_topic(bag, msg_topic, 'sensor_msgs/msg/Imu')
    for timestamp, oxts in zip(kitti.timestamps, kitti.oxts):
        q = transformations.quaternion_from_euler(oxts.packet.roll, oxts.packet.pitch, oxts.packet.yaw)
        
        imu = Imu()
        imu.header.frame_id = imu_frame_id
        
        #imu.header.stamp = rospy.Time.from_sec(float(timestamp.strftime("%s.%f")))
        Timer = rospy.time.Time(seconds = int(timestamp.strftime("%s")), nanoseconds = int(timestamp.strftime("%f"))*10e7)
        imu.header.stamp = Timer.to_msg()
        imu.orientation.x = q[0]
        imu.orientation.y = q[1]
        imu.orientation.z = q[2]
        imu.orientation.w = q[3]
        imu.linear_acceleration.x = oxts.packet.af
        imu.linear_acceleration.y = oxts.packet.al
        imu.linear_acceleration.z = oxts.packet.au
        imu.angular_velocity.x = oxts.packet.wf
        imu.angular_velocity.y = oxts.packet.wl
        imu.angular_velocity.z = oxts.packet.wu
        bag.write(msg_topic, serialize_message(imu), Timer.nanoseconds)


def save_dynamic_tf(bag, kitti, kitti_type, initial_time):
    print("Exporting time dependent transformations")
    msg_topic = '/tf'
    create_topic(bag, msg_topic, 'tf2_msgs/msg/TFMessage')
    if kitti_type.find("raw") != -1:
        for timestamp, oxts in zip(kitti.timestamps, kitti.oxts):
            tf_oxts_msg = TFMessage()
            tf_oxts_transform = TransformStamped()
            Timer = rospy.time.Time(seconds = int(timestamp.strftime("%s")), nanoseconds = int(timestamp.strftime("%f"))*10e7)
            tf_oxts_transform.header.stamp = Timer.to_msg()
            tf_oxts_transform.header.frame_id = 'world'
            tf_oxts_transform.child_frame_id = 'base_link'

            transform = (oxts.T_w_imu)
            t = transform[0:3, 3]
            q = transformations.quaternion_from_matrix(transform)
            oxts_tf = Transform()

            oxts_tf.translation.x = t[0]
            oxts_tf.translation.y = t[1]
            oxts_tf.translation.z = t[2]

            oxts_tf.rotation.x = q[0]
            oxts_tf.rotation.y = q[1]
            oxts_tf.rotation.z = q[2]
            oxts_tf.rotation.w = q[3]

            tf_oxts_transform.transform = oxts_tf
            tf_oxts_msg.transforms.append(tf_oxts_transform)
            
            bag.write(msg_topic, serialize_message(tf_oxts_msg), Timer.nanoseconds)

    elif kitti_type.find("odom") != -1:
        timestamps = map(lambda x: initial_time + x.total_seconds(), kitti.timestamps)
        for timestamp, tf_matrix in zip(timestamps, kitti.T_w_cam0):
            tf_msg = TFMessage()
            tf_stamped = TransformStamped()
            #tf_stamped.header.stamp = rospy.Time.from_sec(timestamp)
            Timer = rospy.time.Time(seconds = int(timestamp.strftime("%s")), nanoseconds = int(timestamp.strftime("%f"))*10e7)
            tf_stamped.header.stamp = Timer.to_msg()
            tf_stamped.header.frame_id = 'world'
            tf_stamped.child_frame_id = 'camera_left'
            
            t = tf_matrix[0:3, 3]
            q = transformations.quaternion_from_matrix(tf_matrix)
            transform = Transform()

            transform.translation.x = t[0]
            transform.translation.y = t[1]
            transform.translation.z = t[2]

            transform.rotation.x = q[0]
            transform.rotation.y = q[1]
            transform.rotation.z = q[2]
            transform.rotation.w = q[3]

            tf_stamped.transform = transform
            tf_msg.transforms.append(tf_stamped)

            bag.write(msg_topic, serialize_message(tf_msg), Timer.nanoseconds)
          
        
def save_camera_data(bag, kitti_type, kitti, util, bridge, camera, camera_frame_id, topic, initial_time):
    print("Exporting camera {}".format(camera))
    
    if kitti_type.find("raw") != -1:
        camera_pad = '{0:02d}'.format(camera)
        image_dir = os.path.join(kitti.data_path, 'image_{}'.format(camera_pad))
        image_path = os.path.join(image_dir, 'data')
        image_filenames = sorted(os.listdir(image_path))
        with open(os.path.join(image_dir, 'timestamps.txt')) as f:
            image_datetimes = map(lambda x: datetime.strptime(x[:-4], '%Y-%m-%d %H:%M:%S.%f'), f.readlines())
        
        calib = CameraInfo()
        calib.header.frame_id = camera_frame_id
        width, height = tuple(util['S_rect_{}'.format(camera_pad)].tolist())
        calib.width = int(width)
        calib.height = int(height)
        calib.distortion_model = 'plumb_bob'
        calib.k = util['K_{}'.format(camera_pad)]
        calib.r = util['R_rect_{}'.format(camera_pad)]
        #print(util['D_{}'.format(camera_pad)])
        calib.d = set(util['D_{}'.format(camera_pad)])
        calib.p = util['P_rect_{}'.format(camera_pad)]
        topic_ext = "/image_raw"
    elif kitti_type.find("odom") != -1:
        camera_pad = '{0:01d}'.format(camera)
        image_path = os.path.join(kitti.sequence_path, 'image_{}'.format(camera_pad))
        image_filenames = sorted(os.listdir(image_path))
        image_datetimes = map(lambda x: initial_time + x.total_seconds(), kitti.timestamps)
        
        calib = CameraInfo()
        calib.header.frame_id = camera_frame_id
        calib.p = util['P{}'.format(camera_pad)]
        topic_ext = "/image_rect"

    img_topic = topic + topic_ext
    calib_topic = topic + '/camera_info'
    create_topic(bag, img_topic, 'cv_bridge/CVBridge/imgmsg')
    create_topic(bag, calib_topic, 'sensor_msgs/msg/CameraInfo')

    iterable = zip(image_datetimes, image_filenames)
    bar = progressbar.ProgressBar()
    for dt, filename in bar(iterable):
        image_filename = os.path.join(image_path, filename)
        cv_image = cv2.imread(image_filename)
        calib.height, calib.width = cv_image.shape[:2]
        if camera in (0, 1):
            cv_image = cv2.cvtColor(cv_image, cv2.COLOR_BGR2GRAY)
        encoding = "mono8" if camera in (0, 1) else "bgr8"
        image_message = bridge.cv2_to_imgmsg(cv_image, encoding=encoding)
        image_message.header.frame_id = camera_frame_id

        if kitti_type.find("raw") != -1:
            #image_message.header.stamp = rospy.Time.from_sec(float(datetime.strftime(dt, "%s.%f")))
            Timer = rospy.time.Time(seconds = int(datetime.strftime(dt, "%s")), nanoseconds = int(datetime.strftime(dt, "%f"))*10e7)
            
        elif kitti_type.find("odom") != -1:
            #image_message.header.stamp = rospy.Time.from_sec(dt)
            Timer = rospy.time.Time(seconds = dt)

    
        image_message.header.stamp = Timer.to_msg()
        calib.header.stamp = image_message.header.stamp
        bag.write(img_topic, serialize_message(image_message), Timer.nanoseconds)
        bag.write(calib_topic, serialize_message(calib), Timer.nanoseconds) 
 
def point_field(pf, name, offset, datatype, count):
    pf.name = name
    pf.offset = offset
    pf.datatype = datatype
    pf.count = count
    return pf

def save_velo_data(bag, kitti, velo_frame_id, topic):
    print("Exporting velodyne data")
    velo_path = os.path.join(kitti.data_path, 'velodyne_points')
    velo_data_dir = os.path.join(velo_path, 'data')
    velo_filenames = sorted(os.listdir(velo_data_dir))
    with open(os.path.join(velo_path, 'timestamps.txt')) as f:
        lines = f.readlines()
        velo_datetimes = []
        for line in lines:
            if len(line) == 1:
                continue
            dt = datetime.strptime(line[:-4], '%Y-%m-%d %H:%M:%S.%f')
            velo_datetimes.append(dt)

    iterable = zip(velo_datetimes, velo_filenames)
    bar = progressbar.ProgressBar()
    msg_topic = topic + '/pointcloud'
    create_topic(bag, msg_topic, 'sensor_msgs_py/point_cloud2/pcl2')
    for dt, filename in bar(iterable):
        if dt is None:
            continue

        velo_filename = os.path.join(velo_data_dir, filename)

        # read binary data
        scan = (np.fromfile(velo_filename, dtype=np.float32)).reshape(-1, 4)

        # create header
        header = Header()
        header.frame_id = velo_frame_id
        #header.stamp = rospy.Time.from_sec(float(datetime.strftime(dt, "%s.%f")))
        Timer = rospy.time.Time(seconds = int(datetime.strftime(dt, "%s")), nanoseconds = int(datetime.strftime(dt, "%f"))*10e7)
        header.stamp = Timer.to_msg()
        # fill pcl msg
        """fields = [PointField('x', 0, PointField.FLOAT32, 1),
                  PointField('y', 4, PointField.FLOAT32, 1),
                  PointField('z', 8, PointField.FLOAT32, 1),
                  PointField('i', 12, PointField.FLOAT32, 1)]"""
        
        
        fields = [point_field(PointField(),'x', 0, PointField.FLOAT32, 1),
                  point_field(PointField(),'y', 4, PointField.FLOAT32, 1),
                  point_field(PointField(),'z', 8, PointField.FLOAT32, 1),
                  point_field(PointField(),'i', 12, PointField.FLOAT32, 1)]

        pcl_msg = pcl2.create_cloud(header, fields, scan)

        bag.write(msg_topic, serialize_message(pcl_msg), Timer.nanoseconds)


def get_static_transform(from_frame_id, to_frame_id, transform):
    t = transform[0:3, 3]
    q = transformations.quaternion_from_matrix(transform)
    tf_msg = TransformStamped()
    tf_msg.header.frame_id = from_frame_id
    tf_msg.child_frame_id = to_frame_id
    tf_msg.transform.translation.x = float(t[0])
    tf_msg.transform.translation.y = float(t[1])
    tf_msg.transform.translation.z = float(t[2])
    tf_msg.transform.rotation.x = float(q[0])
    tf_msg.transform.rotation.y = float(q[1])
    tf_msg.transform.rotation.z = float(q[2])
    tf_msg.transform.rotation.w = float(q[3])
    return tf_msg


def inv(transform):
    "Invert rigid body transformation matrix"
    R = transform[0:3, 0:3]
    t = transform[0:3, 3]
    t_inv = -1 * R.T.dot(t)
    transform_inv = np.eye(4)
    transform_inv[0:3, 0:3] = R.T
    transform_inv[0:3, 3] = t_inv
    return transform_inv


def save_static_transforms(bag, transforms, timestamps):
    print("Exporting static transformations")
    msg_topic = '/tf_static'
    create_topic(bag, msg_topic, 'tf2_msgs/msg/TFMessage')
    tfm = TFMessage()
    for transform in transforms:
        t = get_static_transform(from_frame_id=transform[0], to_frame_id=transform[1], transform=transform[2])
        tfm.transforms.append(t)
    for timestamp in timestamps:
        #time = rospy.Time.from_sec(float(timestamp.strftime("%s.%f")))
        Timer = rospy.time.Time(seconds = int(timestamp.strftime("%s")), nanoseconds = int(timestamp.strftime("%f"))*10e7)
        time = Timer.to_msg()
        for i in range(len(tfm.transforms)):
            tfm.transforms[i].header.stamp = time
        bag.write(msg_topic, serialize_message(tfm), Timer.nanoseconds)


def save_gps_fix_data(bag, kitti, gps_frame_id, topic):
    msg_topic = topic
    create_topic(bag, msg_topic, 'sensor_msgs/msg/NavSatFix')
    for timestamp, oxts in zip(kitti.timestamps, kitti.oxts):
        navsatfix_msg = NavSatFix()
        navsatfix_msg.header.frame_id = gps_frame_id
        #navsatfix_msg.header.stamp = rospy.Time.from_sec(float(timestamp.strftime("%s.%f")))
        Timer = rospy.time.Time(seconds = int(timestamp.strftime("%s")), nanoseconds = int(timestamp.strftime("%f"))*10e7)
        navsatfix_msg.header.stamp = Timer.to_msg()
        navsatfix_msg.latitude = oxts.packet.lat
        navsatfix_msg.longitude = oxts.packet.lon
        navsatfix_msg.altitude = oxts.packet.alt
        navsatfix_msg.status.service = 1
        
        
        bag.write(msg_topic, serialize_message(navsatfix_msg), Timer.nanoseconds)


def save_gps_vel_data(bag, kitti, gps_frame_id, topic):
    msg_topic = topic
    create_topic(bag, msg_topic, 'geometry_msgs/msg/TwistStamped')
    for timestamp, oxts in zip(kitti.timestamps, kitti.oxts):
        twist_msg = TwistStamped()
        twist_msg.header.frame_id = gps_frame_id
        #twist_msg.header.stamp = rospy.Time.from_sec(float(timestamp.strftime("%s.%f")))
        Timer = rospy.time.Time(seconds = int(timestamp.strftime("%s")), nanoseconds = int(timestamp.strftime("%f"))*10e7)
        twist_msg.header.stamp = Timer.to_msg()
        twist_msg.twist.linear.x = oxts.packet.vf
        twist_msg.twist.linear.y = oxts.packet.vl
        twist_msg.twist.linear.z = oxts.packet.vu
        twist_msg.twist.angular.x = oxts.packet.wf
        twist_msg.twist.angular.y = oxts.packet.wl
        twist_msg.twist.angular.z = oxts.packet.wu
        bag.write(msg_topic, serialize_message(twist_msg), Timer.nanoseconds)



def run_kitti2bag2():
    parser = argparse.ArgumentParser(description = "Convert KITTI dataset to ROS bag file the easy way!")
    # Accepted argument values
    kitti_types = ["raw_synced", "odom_color", "odom_gray"]
    odometry_sequences = []
    for s in range(22):
        odometry_sequences.append(str(s).zfill(2))
    
    parser.add_argument("kitti_type", choices = kitti_types, help = "KITTI dataset type")
    parser.add_argument("dir", nargs = "?", default = os.getcwd(), help = "base directory of the dataset, if no directory passed the deafult is current working directory")
    parser.add_argument("-t", "--date", help = "date of the raw dataset (i.e. 2011_09_26), option is only for RAW datasets.")
    parser.add_argument("-r", "--drive", help = "drive number of the raw dataset (i.e. 0001), option is only for RAW datasets.")
    parser.add_argument("-s", "--sequence", choices = odometry_sequences,help = "sequence of the odometry dataset (between 00 - 21), option is only for ODOMETRY datasets.")
    args = parser.parse_args()

    bridge = CvBridge()
    #compression = rosbag.CompressionMode.NONE
    # compression = rosbag.Compression.BZ2
    # compression = rosbag.Compression.LZ4
    
    # CAMERAS
    cameras = [
        (0, 'camera_gray_left', '/kitti/camera_gray_left'),
        (1, 'camera_gray_right', '/kitti/camera_gray_right'),
        (2, 'camera_color_left', '/kitti/camera_color_left'),
        (3, 'camera_color_right', '/kitti/camera_color_right')
    ]

    if args.kitti_type.find("raw") != -1:
    
        if args.date == None:
            print("Date option is not given. It is mandatory for raw dataset.")
            print("Usage for raw dataset: kitti2bag raw_synced [dir] -t <date> -r <drive>")
            sys.exit(1)
        elif args.drive == None:
            print("Drive option is not given. It is mandatory for raw dataset.")
            print("Usage for raw dataset: kitti2bag raw_synced [dir] -t <date> -r <drive>")
            sys.exit(1)
        tmp_path = 'temp'
        bag_path = str(tmp_path + '/tmp_write_test')

        storage_options, converter_options = get_rosbag_options(bag_path)
        bag = rosbag2_py.SequentialWriter()
        bag.open(storage_options, converter_options)
        kitti = pykitti.raw(args.dir, args.date, args.drive)

        if not os.path.exists(kitti.data_path):
            print('Path {} does not exists. Exiting.'.format(kitti.data_path))
            sys.exit(1)

        if len(kitti.timestamps) == 0:
            print('Dataset is empty? Exiting.')
            sys.exit(1)

        try:
            # IMU
            imu_frame_id = 'imu_link'
            imu_topic = '/kitti/oxts/imu'
            gps_fix_topic = '/kitti/oxts/gps/fix'
            gps_vel_topic = '/kitti/oxts/gps/vel'
            velo_frame_id = 'velo_link'
            velo_topic = '/kitti/velo'

            T_base_link_to_imu = np.eye(4, 4)
            T_base_link_to_imu[0:3, 3] = [-2.71/2.0-0.05, 0.32, 0.93]

            # tf_static
            transforms = [
                ('base_link', imu_frame_id, T_base_link_to_imu),
                (imu_frame_id, velo_frame_id, inv(kitti.calib.T_velo_imu)),
                (imu_frame_id, cameras[0][1], inv(kitti.calib.T_cam0_imu)),
                (imu_frame_id, cameras[1][1], inv(kitti.calib.T_cam1_imu)),
                (imu_frame_id, cameras[2][1], inv(kitti.calib.T_cam2_imu)),
                (imu_frame_id, cameras[3][1], inv(kitti.calib.T_cam3_imu))
            ]

            util = pykitti.utils.read_calib_file(os.path.join(kitti.calib_path, 'calib_cam_to_cam.txt'))


            save_static_transforms(bag, transforms, kitti.timestamps)
            save_dynamic_tf(bag, kitti, args.kitti_type, initial_time=None)
            save_imu_data(bag, kitti, imu_frame_id, imu_topic)
            save_gps_fix_data(bag, kitti, imu_frame_id, gps_fix_topic)
            save_gps_vel_data(bag, kitti, imu_frame_id, gps_vel_topic)
            for camera in cameras:
                save_camera_data(bag, args.kitti_type, kitti, util, bridge, camera=camera[0], camera_frame_id=camera[1], topic=camera[2], initial_time=None)
            save_velo_data(bag, kitti, velo_frame_id, velo_topic)

        finally:
            print("## OVERVIEW ##")
            print(bag)
            del bag
```

I have successfully recorded the data. 
But when I tried publishing the bags using the following commands.

```bash
source ~/ros2_rolling/install/setup.bash && source ~/ros_pkgs/install/setup.bash
ros2 bag play temp/tmp_write_test
```

It is publishing normal topics of std_msg. But is ignoring a few topics by showing the following warning.

```bash
[INFO] [1613380015.893647871] [rosbag2_storage]: Opened database 'temp/tmp_write_test/tmp_write_test_0.db3' for READ_ONLY.
[WARN] [1613380015.899220038] [rosbag2_transport]: Ignoring a topic '/kitti/camera_gray_right/image_raw', reason: dlopen error: /home/ubuntu/ros_pkgs/install/cv_bridge/lib/libcv_bridge__rosidl_typesupport_cpp.so: cannot open shared object file: No such file or directory, at /home/ubuntu/ros2_rolling/src/ros2/rcutils/src/shared_library.c:99.
[WARN] [1613380015.903106174] [rosbag2_transport]: Ignoring a topic '/kitti/camera_gray_left/image_raw', reason: dlopen error: /home/ubuntu/ros_pkgs/install/cv_bridge/lib/libcv_bridge__rosidl_typesupport_cpp.so: cannot open shared object file: No such file or directory, at /home/ubuntu/ros2_rolling/src/ros2/rcutils/src/shared_library.c:99.
[WARN] [1613380015.907897351] [rosbag2_transport]: Ignoring a topic '/kitti/velo/pointcloud', reason: dlopen error: /home/ubuntu/ros2_rolling/install/sensor_msgs_py/lib/libsensor_msgs_py__rosidl_typesupport_cpp.so: cannot open shared object file: No such file or directory, at /home/ubuntu/ros2_rolling/src/ros2/rcutils/src/shared_library.c:99.
[WARN] [1613380015.910327644] [rosbag2_transport]: Ignoring a topic '/kitti/camera_color_left/image_raw', reason: dlopen error: /home/ubuntu/ros_pkgs/install/cv_bridge/lib/libcv_bridge__rosidl_typesupport_cpp.so: cannot open shared object file: No such file or directory, at /home/ubuntu/ros2_rolling/src/ros2/rcutils/src/shared_library.c:99.
[WARN] [1613380015.912786349] [rosbag2_transport]: Ignoring a topic '/kitti/camera_color_right/image_raw', reason: dlopen error: /home/ubuntu/ros_pkgs/install/cv_bridge/lib/libcv_bridge__rosidl_typesupport_cpp.so: cannot open shared object file: No such file or directory, at /home/ubuntu/ros2_rolling/src/ros2/rcutils/src/shared_library.c:99.
```

I have cv_bridge package in ~/ros_pkgs workspace. And ros2 is built in ~/ros2_rolling and sourced both of them while building this package. And I pulled latest source and tested on it just before writing this issue.

## System (please complete the following information)
 - OS:  Ubuntu Focal
 - ROS 2 Distro: Rolling
 - Version: master

**Comment by emersonknapp** [comment]
I'm just going to note that this repro code is _too long to read_ - are you able to provide a more minimal reproduction case? E.g. `ros2 topic pub cv_msgs/Whatever` piped into a rosbag2?

To address the warnings - do the mentioned files such as `/home/ubuntu/ros_pkgs/install/cv_bridge/lib/libcv_bridge__rosidl_typesupport_cpp.so` actually exist? If they are not there, are they somewhere else? (`find /home/ubuntu -name '*libcv_bridge*'`)

**Comment by jaan1729** [comment]
Thanks for the responce. I tried to publish the fake data and compared with this code.

I used wrong message types while creating topics. Correcting them helped.

Ex:
1.
```python
create_topic(bag, img_topic, 'cv_bridge/CVBridge/imgmsg')  #Incorrect
create_topic(bag, img_topic, 'sensor_msgs/msg/Image')  #Correct
```

2.
```python
create_topic(bag, msg_topic, 'sensor_msgs_py/point_cloud2/pcl2')  #Incorrect
create_topic(bag, msg_topic, 'sensor_msgs/msg/PointCloud2')   #Correct
```

The issue is resolved.

**Comment by fabrizioschiano** [comment]
Hi @jaan1729 , did you understand why `sensor_msgs_py` did not work? I m stumbling across the same problem. Thanks!
